### PR TITLE
fix(core): handle schema Cache expandCapacity concurrent problem

### LINKED ISSUE #2330: [Bug] Schema Cache expandCapacity concurrent issue
### Bug Type (问题类型)

other exception / error (其他异常报错)

### Before submit

- [X] 我已经确认现有的 [Issues](https://github.com/apache/hugegraph/issues) 与 [FAQ](https://hugegraph.apache.org/docs/guides/faq/) 中没有相同 / 重复问题 (I have confirmed and searched that there are no similar problems in the historical issue and documents)

### Environment (环境信息)

see the following error-log
```java
org.apache.hugegraph.exception.ServerException: Index 32 out of bounds for length 32
```

### Expected & Actual behavior (期望与实际表现)

```java
    public void set(int key, V value) {
        if (key >= this.currentSize) {
            this.expandCapacity();
        }
        // this.array is not the one that has done expandCapacity!
        this.array[key] = value;
    }

    private synchronized void expandCapacity() {
        if (this.currentSize == this.maxSize) {
            return;
        }
        this.currentSize = Math.min(this.currentSize * 2, this.maxSize);
        Object[] newArray = new Object[this.currentSize];
        System.arraycopy(this.array, 0, newArray, 0, this.array.length);
        this.clear();
        this.array = newArray;
    }
```

### Vertex/Edge example (问题点 / 边数据举例)

_No response_

### Schema [VertexLabel, EdgeLabel, IndexLabel] (元数据结构)

_No response_

### PR DESCRIPTION
**Pull request by conghuhu**
## Purpose of the PR

fix #2330 

## Verifying these changes

- [ ] Trivial rework / code cleanup without any test coverage. (No Need)
- [ ] Already covered by existing tests, such as *(please modify tests here)*.
- [x] Need tests and can be verified as follows:
    - org.apache.hugegraph.unit.store.RamIntObjectMapTest

## Does this PR potentially affect the following parts?

- [x]  Nope
- [ ]  Dependencies (add/update license info) 
- [ ]  Modify configurations
- [ ]  The public API
- [ ]  Other affects (typed here)

## Documentation Status

- [ ]  `Doc - TODO` 
- [x]  `Doc - Done` 
- [ ]  `Doc - No Need`

### CODE REVIEW

**Comment by conghuhu** [comment]
@imbajin PTAL

File: hugegraph-core/src/main/java/org/apache/hugegraph/backend/store/ram/IntObjectMap.java [hunk 1]
```diff
@@ -29,8 +29,8 @@ public final class IntObjectMap<V> implements RamMap {
     private static final float DEFAULT_INITIAL_FACTOR = 0.25f;
 
     private final int maxSize;
-    private int currentSize;
-    private Object[] array;
+    private volatile int currentSize;
+    private volatile Object[] array;
```
Comment (imbajin) [inline hugegraph-core/src/main/java/org/apache/hugegraph/backend/store/ram/IntObjectMap.java:33]: seems we just need mark it as `final` to ensure the init integrity? could try it to ensure
```suggestion
    private final Object[] array;
```

Comment (conghuhu) [reply hugegraph-core/src/main/java/org/apache/hugegraph/backend/store/ram/IntObjectMap.java:33]: image
If it is changed to final, then the value cannot be assigned again

Comment (imbajin) [reply hugegraph-core/src/main/java/org/apache/hugegraph/backend/store/ram/IntObjectMap.java:33]: > image If it is changed to final, then the value cannot be assigned again

emm, if we need to assign multiple times, it cannot be done like this, ignore it

**Comment by imbajin** [comment]
Find another critical bug @conghuhu 

image

**Pull request merged**